core/iwasm/common/wasm_c_api.c

```diff
@@ -335,8 +348,16 @@ wasm_engine_new()
         singleton_engine =
             wasm_engine_new_internal(Alloc_With_System_Allocator, NULL);
     }
-    if (singleton_engine)
+    else {
+        os_mutex_lock(&singleton_engine->lock);
+
+        if (singleton_engine->ref_count == 0)
```

Comment:
i don't understand what this check is for.

isn't it simpler to use a global lock to protect singleton_engine?

Reply:
Yes, using a global lock would be a better way, but not all platforms have the global lock.